frontends/p4/typeChecking/typeChecker.cpp

```diff
@@ -3512,6 +3512,15 @@ const IR::Expression *TypeInference::actionCall(bool inActionList,
                     typeError("%1%: action argument must be a compile-time constant",
                               arg->expression);
             }
+            // This is like an assignment; may make additional conversions.
+            newExpr = assignment(arg, param->type, arg->expression);
+            if (readOnly) {
```

Comment:
Can you elaborate on that bug? Maybe we can fix the DPDK back end.

Reply:
Basically, the DPDK backend creates new code that is not type-correct, so a following TypeChecking pass tries to introduce new implicit casts, but fails as the code is `readOnly`.  I'm not sure why it only fails after this change, as it seems like it should fail regardless.

I tried adding a additional TypeInferencing pass in the DPDK backend to introduce the implicit casts correctly, but that caused some other problems.  This feels like a bit of a hack, leaving the code as semi-type-incorrect (not making the implicit cast explicit) when the readOnly flag is set.

Reply:
Which pass is this? Maybe we can fix the generation of this action call in the first place?

I think this is something the DPDK back end needs to fix before we add more workarounds to the type checker. The class is already complicated.

Reply:
The problem is in DPDK::PrependPDotToActionArgs (line ~1318) where it is creating a ListExpression rather than a StructExpression to initialize a struct.  The typechecker wants this to be a StructExpression, so will introduce a "cast" (actually rewrite the ListExpression as a StructExpression), which then causes a problem as it is running with `readOnly = true` (backend.cpp line 107).  If I change that call to have readOnly = false, then it doesn't assert and just makes the change.  However, it then has problems later in the backend with code that doesn't know what to do with a StructExpression.

Without my change, the typeChecker does not correctly type check action args, so it doesn't try to insert the conversion.

The easiest fix is probably just to add a call to P4::TypeInference with readOnly = false somewhere, and fix the backend code to understand a StructExpression.

Reply:
@usha1830 Could you take a look at this?

Reply:
#4553

Reply:
Sure. I will check this out.